### PR TITLE
refactor hadoop daos

### PR DESCRIPTION
**Pull request by minmingzhu**
1. use latest daos-java-api to implement hadoop daos
2. refactor DaosInputStream/DaosOutputStream

### CODE REVIEW

File: java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/Constants.java [hunk 1]
```diff
@@ -18,14 +18,13 @@
 
 package com.intel.daos.hadoop.fs;
 
+import com.intel.daos.client.DaosObjectType;
```
Comment (gczsjdy) [inline java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/Constants.java]: I think they are too many lines changed in this PR that are not relevant or has no actual effects:
- 1) For example this is never used?
- 2) Style and log-relevant changes can be consolidated to one commit, this way if I want to ignore these changes I can see other commits

**Comment by gczsjdy** [comment]
@minmingzhu @zjf2012 @carsonwang 
FYI, in the meantime you are reviewing this, I am using this version(the 2 commits ) to develop buffered read/non-buffered read(depending on an introduced config entry flag). So let's not make big changes on this.

I will try to test my logics using daos-java-api mocks, instead of depending on real daos-java-api and underlying DAOS. About this part I think we have agreed on the same page?

File: java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosFileSystem.java [hunk 2]
```diff
@@ -59,26 +57,26 @@ public void initialize(URI name, Configuration conf)
...
       String svc = conf.get(Constants.DAOS_POOL_SVC);
-      this.client = new DaosFS.Builder(false, false, false)
-              .setPooluuid(uuid)
-              .setContaineruuid(mountpoint)
-              .setSvc(svc).build();
+      if(svc==null) throw new IOException(Constants.DAOS_POOL_SVC + " is null, need to set up " + Constants.DAOS_POOL_SVC +" in hdfs.xml .");
+
+      // daosFSclient build
+      this.daos = new DaosFsClient.DaosFsClientBuilder().poolId(pooluuid).containerId(contuuid).svc(svc).build();
       this.uri = URI.create(name.getScheme() + "://" + name.getAuthority());
```
Comment (zjf2012) [inline java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosFileSystem.java]: We don't use daos://ip:port to connect to daos server. Thus, we can remove this info.

File: java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosFileSystem.java [hunk 3]
```diff
@@ -37,19 +39,15 @@
  * used to access DAOS blob system in a filesystem style.
  */
 public class DaosFileSystem extends FileSystem {
```
Comment (zjf2012) [inline java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosFileSystem.java]: For checking if file is directory or existed, we don't need to depend on hdfs' FileStatus which should be provided only for public HDFS APIs.

We should try to instantiate DaosFile object as earlier as possible and avoid multiple DaosFile instances as much as we can.

File: java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosFileSystem.java [hunk 4]
```diff
@@ -139,38 +125,24 @@ public FSDataOutputStream create(Path f,
...
     }
     String key = f.toUri().getPath();
-    if(!exists(f.getParent())){
-      try {
-        mkdirs(f.getParent(), new FsPermission("755"));
-      }catch (IOException e){
-
-      }
-    }
     if (exists(f)) {
```
Comment (zjf2012) [inline java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosFileSystem.java]: We should try to instantiate DaosFile object as earlier as possible and avoid multiple DaosFile instances as much as we can.

File: java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosFileSystem.java [hunk 5]
```diff
@@ -183,34 +155,42 @@ public FSDataOutputStream append(Path f,
...
-    int rc = -1;
+    // determine  if src is root dir and whether it exits
     if(src.toUri().getPath().equals("/")){
+      if(LOG.isDebugEnabled()){
+        LOG.debug("DaosFileSystem:  cat not rename root path %s",src);
+      }
+      return false;
+    }
+
+    FileStatus srcStatus = null;
```
Comment (zjf2012) [inline java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosFileSystem.java]: we don't need FileStatus

File: java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosFileSystem.java [hunk 6]
```diff
@@ -356,21 +377,22 @@ public FileStatus getFileStatus(Path f) throws IOException {
...
+    } catch (IOException e) {
       return false;
     }
   }
 
   @Override
   public void close() throws IOException {
     if (LOG.isDebugEnabled()) {
-      LOG.debug("close");
+      LOG.debug("DaosFileSystem close");
```
Comment (zjf2012) [inline java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosFileSystem.java]: DaosFsClient.disconnect should be called here

File: java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosInputStream.java [hunk 7]
```diff
@@ -160,137 +167,112 @@ public synchronized int read() throws IOException {
   @Override
   public synchronized int read(byte[] buf, int off, int len)
```
Comment (zjf2012) [inline java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosInputStream.java]: add logic to cover case that the buf array has longer length than internal buffer

File: java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosOutputStream.java [hunk 8]
```diff
@@ -58,13 +57,16 @@ public DaosOutputStream(DaosFile daosFile,
    */
   @Override
   public synchronized void write(int b) throws IOException {
```
Comment (zjf2012) [inline java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosOutputStream.java]: add logic to cover case that the buf array has longer length than internal buffer

File: java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosFileSystem.java (hunk 2 above)
Comment (minmingzhu) [reply java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosFileSystem.java]: we can don't use daos://ip:port ,  The reason why this is used here is to correspond to hdfs, s3, and aliyun .

File: java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosInputStream.java (hunk 7 above)
Comment (minmingzhu) [reply java/hadoop-daos/src/main/java/com/intel/daos/hadoop/fs/DaosInputStream.java]: If the length of the buf array is longer length than the cache buffer, a circular read will be performed . you can reference to following code . 
```
 int offset = this.buffer.remaining();
        this.buffer.put(b,off,this.buffer.remaining());
        if (!this.buffer.hasRemaining()) {
          daoswrite();
        }
        write(b,offset,len);
```

**Pull request merged**